src/main/src/main/java/org/geoserver/ows/DisabledServiceCheck.java

```diff
@@ -81,6 +112,52 @@ public Service serviceDispatched(Request request, Service service) {
         return service;
     }
 
+    /**
+     * Extracts the context from the request, used to check for layer reference.
+     *
+     * @param request The request.
+     * @return The context.
+     */
+    String context(Request request) {
+        String context = request.getContext();
+        if (context != null && context.contains("gwc/service")) {
+            // Account for context provided to GwcServiceDispatcherCallback
+            if (context.contains("/gwc/service")) {
+                context = context.substring(0, context.indexOf("/gwc/service"));
+            } else {
+                context = context.substring(0, context.indexOf("gwc/service"));
+            }
+            // remove leading and trailing slashes
+            if (context.startsWith("/")) {
+                context = context.substring(1);
+            }
+            if (context.endsWith("/")) {
+                context = context.substring(0, context.length() - 1);
+            }
+            int idx1 = context.indexOf("/");
+            if (idx1 != -1) {
+                int idx2 = context.indexOf("/", idx1 + 1);
+                if (idx2 != -1) {
+                    return context.substring(0, idx2);
+                }
+            }
+        }
+        return context;
+    }
+
+    LayerInfo getLayerByName(String layerName) {
+        // We need access to actual catalog, not filtered for current service
+        Catalog catalog = geoServer.getCatalog();
+        while (catalog instanceof Wrapper && ((Wrapper) catalog).isWrapperFor(Catalog.class)) {
+            Catalog unwrapped = ((Wrapper) catalog).unwrap(Catalog.class);
+            if (unwrapped == catalog || unwrapped == null) {
+                break;
+            }
+            catalog = unwrapped;
+        }
+        return catalog.getLayerByName(layerName);
```

Comment:
Confused... when is catalog filtered "by service"'?

Reply:
When providing WMS then all the layers that have been disabled for WMS need to be filtered. 

This is the issue that is being addressed, producing a layer WMS that has no content. Rather than a 404.